test/integration/integration.cc

```diff
@@ -414,10 +414,25 @@ void BaseIntegrationTest::setUpstreamAddress(
   socket_address->set_port_value(fake_upstreams_[upstream_index]->localAddress()->ip()->port());
 }
 
+namespace {
+bool read_flag(bool* flag) { return *flag; }
+} // namespace
+
 void BaseIntegrationTest::registerTestServerPorts(const std::vector<std::string>& port_names) {
-  auto port_it = port_names.cbegin();
-  auto listeners = test_server_->server().listenerManager().listeners();
+  bool listeners_ready = false;
+  absl::Mutex l;
+  std::vector<std::reference_wrapper<Network::ListenerConfig>> listeners;
+  test_server_->server().dispatcher().post([this, &listeners, &listeners_ready, &l]() {
+    listeners = test_server_->server().listenerManager().listeners();
+    l.Lock();
+    listeners_ready = true;
+    l.Unlock();
+  });
+  l.LockWhen(absl::Condition(read_flag, &listeners_ready));
```

Comment:
Should a `CondVar` be used here, rather than manually doing this with a bool?

Reply:
Interesting... My understanding is that CondVar doesn't maintain the bool flag in this scenario.
It's always 3 components: Mutex, CondVar and Flag
and the core is 
```
while (flag) {
   cond_var.wait(mutex);
}
```
Not sure how we could save the bool flag...

Reply:
While looking up CondVar I accidentally find read_flag is a overhead. Removed